base/linalg/eigen.jl

```diff
@@ -99,6 +99,7 @@ julia> F[:vectors]
 """
 function eigfact(A::StridedMatrix{T}; permute::Bool=true, scale::Bool=true) where T
     S = promote_type(Float32, typeof(one(T)/norm(one(T))))
+    isdiag(A) && return eigfact(Diagonal(A))
```

Comment:
`eigfact(StridedMatrix)` is already quite type-unstable but better to use `S` in this branch as well.